torch/nn/modules/loss.py

```diff
@@ -1246,6 +1253,144 @@ def forward(self, anchor: Tensor, positive: Tensor, negative: Tensor) -> Tensor:
                                      eps=self.eps, swap=self.swap, reduction=self.reduction)
 
 
+class TripletMarginLossWithDistance(_Loss):
+    r"""Creates a criterion that measures the triplet loss given input
+    tensors :math:`a`, :math:`p`, and :math:`n` (representing anchor,
+    positive, and negative examples, respectively); and a real-valued function
+    between them.
+
+    The unreduced loss (i.e., with `reduction` set to `'none'`)
+    can be described as:
+
+    .. math::
+        \ell(a, p, n) = L = \{l_1,\dots,l_N\}^\top, \quad
+        l_i = \max \{d(a_i, p_i) - d(a_i, n_i) + {\rm margin}, 0\}
+
+    where :math:`N` is the batch size; :math:`d` is a real-valued function quantifying
+    the separation between two tensors, referred to as `distance_function`;
+    and :math:`margin` is a non-negative margin enforced between the positive and
+    negative distances.  The input tensors have :math:`N` elements each and can be of
+    any shape that the distance function can handle.
+
+    If :attr:`reduction` is not ``'none'``
+    (default ``'mean'``), then:
+
+    .. math::
+        \ell(x, y) =
+        \begin{cases}
+            \operatorname{mean}(L), &  \text{if reduction} = \text{`mean';}\\
+            \operatorname{sum}(L),  &  \text{if reduction} = \text{`sum'.}
+        \end{cases}
+
+    See also :class:`~torch.nn.TripletMarginLoss`, which computes the triplet
+    loss for input tensors using the :math:`l_p` distance as the distance function.
+
+    Args:
+        distance_function (callable, optional): A distance function between two Tensors which,
```

Comment:
You use both "similarity" and "distance" to describe this function. What are the constraints? Should it always return something positive? Does it have to be a metric?

Reply:
Perhaps it'd be more accurate to refer to this as a metric function.  Does TripletMarginWithMetricLoss sound odd?

Reply:
It does, but then things like `nn.CosineSimilarity()` that you mentioned above wouldn't fit into it?

Reply:
That's true.  `nn.CosineSimilarity()` is defined under `nn.modules.distance` though.  How about sticking with "distance" for the arg name but clarifying in the docs that the function is really "a real-valued, nonnegative function that quantifies the relationship between two tensors"?